### ISSUE TITLE
Agent Prematurely Terminates Long Running Extensions

### ISSUE THREAD

**Issue by boumenot** [Azure/WALinuxAgent#1270]
A user opened an issue against [Azure/custom-script-extension-linux #139](https://github.com/Azure/custom-script-extension-linux/issues/139) about the extension prematurely timing out.  I have verified this issue against 2.2.29.

```sh
az vm extension set -g my-resource-group --vm-name vm1 --publisher Microsoft.Azure.Extensions --name CustomScript --version 2.0 --settings "{ \"commandToExecute\": \"sleep 360; ls\" }"
```

The agent will kill the process after five minutes, which is different from previous behavior.  The logs for the agent claim that it moved the extension PID into a cGroup, but the PID it referenced in the logs is the PID *agent ext-handler process* not the extension.

**Comment by jasonzio** [comment]
If I remember right, when the _ext-handler_ instance is asked to perform an action on an extension, it forks itself; the child is responsible for invoking and monitoring the extension which actually performs the requested action. The parent _ext-handler_ arranges for the forked child to run in the cgroup associated with the extension, enabling resource consumption by the child (on behalf of the extension it's invoking) to be tracked (and, eventually, to be limited).

In other words, I'm expecting the pid of the process-moved-to-cgroup to be that of an instance of the agent, but it should be a forked child of the _ext-handler_ instance of the agent, not the _ext-handler_ instance itself.

**Comment by jermth** [comment]
Hi, I was the one who filed the first issue referenced above. I've attached the handler and waagent logs.
[northcentralus.handler.log](https://github.com/Azure/WALinuxAgent/files/2220106/northcentralus.handler.log)
[northcentralus.waagent.log](https://github.com/Azure/WALinuxAgent/files/2220107/northcentralus.waagent.log)

I only see this in NC US though, not in the other regions I've tested (eastus, westus2)

**Comment by boumenot** [comment]
Thanks for clarifying @jasonzio. I was unsure on why that PID was the parent, but if that is expected that helps. 

We still need to track down why the process is being terminated prematurely.  CustomScript (and other extensions) play games with timeouts. The agent timeout for enable is five minutes, but the ARM timeout is 90 minutes. CustomScript runs scripts for even longer because it detaches before the five minutes is up, and uses the 90 minute timeout.  For some reason, the agent is still associated even after five minutes and then kills the process.

@jermth - the agent has only been released to West Central US and North Central US, which is why your timeouts are localized to that region.
